trie/zk_trie_node.go

```diff
@@ -129,57 +131,100 @@ func NewEmptyNode() *Node {
 
 // NewNodeFromBytes creates a new node by parsing the input []byte.
 func NewNodeFromBytes(b []byte) (*Node, error) {
+	var n Node
+	if err := n.SetBytes(b); err != nil {
+		return nil, err
+	}
+	return &n, nil
+}
+
+// LeafHash computes the key of a leaf node given the hIndex and hValue of the
+// entry of the leaf.
+func LeafHash(k, v *zkt.Hash) (*zkt.Hash, error) {
+	return zkt.HashElemsWithDomain(big.NewInt(int64(NodeTypeLeaf_New)), k.BigInt(), v.BigInt())
+}
+
+func (n *Node) SetBytes(b []byte) error {
 	if len(b) < 1 {
-		return nil, ErrNodeBytesBadSize
+		return ErrNodeBytesBadSize
 	}
-	n := Node{Type: NodeType(b[0])}
+	nType := NodeType(b[0])
 	b = b[1:]
-	switch n.Type {
+	switch nType {
 	case NodeTypeParent, NodeTypeBranch_0,
 		NodeTypeBranch_1, NodeTypeBranch_2, NodeTypeBranch_3:
 		if len(b) != 2*zkt.HashByteLen {
-			return nil, ErrNodeBytesBadSize
+			return ErrNodeBytesBadSize
+		}
+
+		childL := n.ChildL
+		childR := n.ChildR
+
+		if childL == nil {
+			childL = zkt.NewHashFromBytes(b[:zkt.HashByteLen])
+		} else {
+			childL.SetBytes(b[:zkt.HashByteLen])
+		}
+
+		if childR == nil {
+			childR = zkt.NewHashFromBytes(b[zkt.HashByteLen : zkt.HashByteLen*2])
+		} else {
+			childR.SetBytes(b[zkt.HashByteLen : zkt.HashByteLen*2])
+		}
+
+		*n = Node{
+			Type:   nType,
+			ChildL: childL,
+			ChildR: childR,
 		}
-		n.ChildL = zkt.NewHashFromBytes(b[:zkt.HashByteLen])
-		n.ChildR = zkt.NewHashFromBytes(b[zkt.HashByteLen : zkt.HashByteLen*2])
 	case NodeTypeLeaf, NodeTypeLeaf_New:
 		if len(b) < zkt.HashByteLen+4 {
-			return nil, ErrNodeBytesBadSize
+			return ErrNodeBytesBadSize
 		}
-		n.NodeKey = zkt.NewHashFromBytes(b[0:zkt.HashByteLen])
+		nodeKey := zkt.NewHashFromBytes(b[0:zkt.HashByteLen])
 		mark := binary.LittleEndian.Uint32(b[zkt.HashByteLen : zkt.HashByteLen+4])
 		preimageLen := int(mark & 255)
-		n.CompressedFlags = mark >> 8
-		n.ValuePreimage = make([]zkt.Byte32, preimageLen)
+		compressedFlags := mark >> 8
+		valuePreimage := slices.Grow(n.ValuePreimage[0:], preimageLen)
```

Comment:
can upgrade the go version to 1.21 now. the upgrade issue encountered before have resolved

Reply:
We would need to upgrade l2geth first, I would rather not do that right now.

Reply:
Colin upgrade this in this pr : https://github.com/scroll-tech/go-ethereum/pull/726/files

Reply:
Ah, nice 👍  I will upgrade zktrie to 1.21 as well.

Reply:
Done